### PR TITLE
feat: implement `wait()` and `kill()` properly

### LINKED ISSUE #10: Bug: Failed to kill pods
```
Events:
  Type     Reason         Age                 From               Message
  ----     ------         ----                ----               -------
  Normal   Scheduled      18m                 default-scheduler  Successfully assigned default/wasm-spin-75bfcd96c-f6m2d to wasm-19508-md-0-628jb
  Normal   Pulling        18m                 kubelet            Pulling image "docker.io/mossaka/spindemo:0.2"
  Normal   Pulled         18m                 kubelet            Successfully pulled image "docker.io/mossaka/spindemo:0.2" in 893.520594ms
  Normal   Created        18m                 kubelet            Created container testwasm
  Normal   Started        18m                 kubelet            Started container testwasm
  Warning  FailedKillPod  8m40s               kubelet            error killing pod: [failed to "KillContainer" for "testwasm" with KillContainerError: "rpc error: code = InvalidArgument desc = failed to stop container \"37c1b531b6f245d7d7eb916f71c856d37a16da5f865126a7be401acd905c171a\": only SIGKILL is supported: invalid argument", failed to "KillPodSandbox" for "33546130-d84b-479c-b8c5-94d795dcae82" with KillPodSandboxError: "rpc error: code = DeadlineExceeded desc = failed to stop container \"37c1b531b6f245d7d7eb916f71c856d37a16da5f865126a7be401acd905c171a\": an error occurs during waiting for container \"37c1b531b6f245d7d7eb916f71c856d37a16da5f865126a7be401acd905c171a\" to be killed: wait container \"37c1b531b6f245d7d7eb916f71c856d37a16da5f865126a7be401acd905c171a\": context deadline exceeded"]
  Warning  FailedKillPod  4m10s               kubelet            error killing pod: [failed to "KillContainer" for "testwasm" with KillContainerError: "rpc error: code = DeadlineExceeded desc = context deadline exceeded", failed to "KillPodSandbox" for "33546130-d84b-479c-b8c5-94d795dcae82" with KillPodSandboxError: "rpc error: code = DeadlineExceeded desc = context deadline exceeded"]
```

When I deployed the spin-shim `v0.1.3` to k8s, and tried to delete the pods, it shows the error msg as above. I believe this is due to a bug in handling kill signal in the shim.

### PR DESCRIPTION
**Pull request by Mossaka**
This PR implements the `wait()` and `kill()` instance functions in Spin Shim. It fixes #10 issue.

Signed-off-by: jiaxiao zhou <jiazho@microsoft.com>

### CODE REVIEW

File: containerd-shim-spin-v1/src/main.rs [hunk 1]
```diff
@@ -126,17 +132,20 @@ impl Instance for Wasi {
...
             bundle: cfg.get_bundle().unwrap_or_default(),
+            shutdown_tx: Arc::new(Mutex::new(tx)),
+            shutdown_rx: Arc::new(Mutex::new(rx)),
         }
     }
     fn start(&self) -> Result<u32, Error> {
         let engine = self.engine.clone();
 
-        let _exit_code = self.exit_code.clone();
+        let exi = self.exit_code.clone();
```
Comment (danbugs) [inline containerd-shim-spin-v1/src/main.rs]: I think this is a typo~ Other than that, LGTM!

**Review by danbugs — APPROVED** [review]

File: containerd-shim-spin-v1/src/main.rs [hunk 2]
```diff
@@ -126,17 +132,20 @@ impl Instance for Wasi {
             stdout: cfg.get_stdout().unwrap_or_default(),
             stderr: cfg.get_stderr().unwrap_or_default(),
             bundle: cfg.get_bundle().unwrap_or_default(),
+            shutdown_tx: Arc::new(Mutex::new(tx)),
```
Comment (cpuguy83) [inline containerd-shim-spin-v1/src/main.rs]: I don't think you need arc/mutex around channel?

Comment (Mossaka) [reply containerd-shim-spin-v1/src/main.rs]: The channel is sent to a thread, so I need it to impl Send and Sync traits. Thus I used Mutex on it. I think I can remove Arc

Comment (cpuguy83) [reply containerd-shim-spin-v1/src/main.rs]: The reason a channel exists is to send data between threads.
I believe all you should need to do is pass a cloned channel end (and not reference the clone anywhere but after moving to the thread).

Comment (Mossaka) [reply containerd-shim-spin-v1/src/main.rs]: It makes sense! Thanks, I will change that.

Comment (Mossaka) [reply containerd-shim-spin-v1/src/main.rs]: rust channel Receiver cannot be `cloned`. Also see here: https://www.reddit.com/r/rust/comments/2lma84/why_cant_you_clone_a_receivert_like_you_can_a/. 

I created the channel in the `new` function and pass the receiver to another thread to wait for the shutdown signal - and this is why I need the `Receiver` to be `Send + Sync`.

Comment (Mossaka) [reply containerd-shim-spin-v1/src/main.rs]: I actually changed the message passing implementation to using `condvar`, which it makes a bit more sense in this case. I only need a `kill` signal to shutdown my server.

File: containerd-shim-spin-v1/src/main.rs [hunk 3]
```diff
@@ -118,6 +122,7 @@ impl Instance for Wasi {
     type E = spin_engine::Engine;
     fn new(id: String, cfg: Option<&InstanceConfig<Self::E>>) -> Self {
         let cfg = cfg.unwrap();
+        let (tx, rx) = std::sync::mpsc::channel::<()>();
```
Comment (cpuguy83) [inline containerd-shim-spin-v1/src/main.rs]: I don't think these are used now?

**Review by cpuguy83 — APPROVED** [review]
One minor issue but otherwise LGTM

**Pull request merged**